value/value.go

```diff
@@ -17,76 +17,170 @@ limitations under the License.
 package value
 
 import (
+	"bytes"
 	"fmt"
-	"sort"
+	"io"
 	"strings"
+
+	jsoniter "github.com/json-iterator/go"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	readPool  = jsoniter.NewIterator(jsoniter.ConfigCompatibleWithStandardLibrary).Pool()
+	writePool = jsoniter.NewStream(jsoniter.ConfigCompatibleWithStandardLibrary, nil, 1024).Pool()
 )
 
-// A Value is an object; it corresponds to an 'atom' in the schema.
-type Value struct {
-	// Exactly one of the below must be set.
-	FloatValue   *Float
-	IntValue     *Int
-	StringValue  *String
-	BooleanValue *Boolean
-	ListValue    *List
-	MapValue     *Map
-	Null         bool // represents an explicit `"foo" = null`
+// FromJSON is a helper function for reading a JSON document.
+func FromJSON(input []byte) (Value, error) {
+	return FromJSONFast(input)
+}
+
+// FromJSONFast is a helper function for reading a JSON document.
+func FromJSONFast(input []byte) (Value, error) {
+	iter := readPool.BorrowIterator(input)
+	defer readPool.ReturnIterator(iter)
+	return ReadJSONIter(iter)
+}
+
+// ToJSON is a helper function for producing a JSon document.
+func ToJSON(v Value) ([]byte, error) {
+	buf := bytes.Buffer{}
+	stream := writePool.BorrowStream(&buf)
+	defer writePool.ReturnStream(stream)
+	WriteJSONStream(v, stream)
+	b := stream.Buffer()
+	err := stream.Flush()
+	// Help jsoniter manage its buffers--without this, the next
+	// use of the stream is likely to require an allocation. Look
+	// at the jsoniter stream code to understand why. They were probably
+	// optimizing for folks using the buffer directly.
+	stream.SetBuffer(b[:0])
+	return buf.Bytes(), err
+}
+
+// ReadJSONIter reads a Value from a JSON iterator.
+func ReadJSONIter(iter *jsoniter.Iterator) (Value, error) {
+	v := iter.Read()
+	if iter.Error != nil && iter.Error != io.EOF {
+		return nil, iter.Error
+	}
+	return NewValueInterface(v), nil
+}
+
+// WriteJSONStream writes a value into a JSON stream.
+func WriteJSONStream(v Value, stream *jsoniter.Stream) {
+	stream.WriteVal(v.Unstructured())
+}
+
+// ToYAML marshals a value as YAML.
+func ToYAML(v Value) ([]byte, error) {
+	return yaml.Marshal(v.Unstructured())
+}
+
+// A Value corresponds to an 'atom' in the schema.
+type Value interface {
+	// IsMap returns true if the Value can be converted to a Map,
+	// false otherwise.
+	IsMap() bool
+	// IsList returns true if the Value can be converted to a List,
+	// false otherwise.
+	IsList() bool
+	// IsBool returns true if the Value can be converted to a bool,
+	// false otherwise.
+	IsBool() bool
+	// IsInt returns true if the Value can be converted to a int64,
+	// false otherwise.
+	IsInt() bool
+	// IsFloat returns true if the Value can be converted to a
+	// float64, false otherwise.
+	IsFloat() bool
+	// IsString returns true if the Value can be converted to a
+	// string, false otherwise.
+	IsString() bool
+	// IsMap returns true if the Value is null, false otherwise.
```

Comment:
IsNull

Make it clear in the comments somewhere that it's a requirement for implementors that exactly one of the IsXXX functions must return true at all times. The schema permits the concept of "a null map" but the value type doesn't-- a thing is either null or a map, never both at the same time.

Reply:
> Make it clear in the comments somewhere that it's a requirement for implementors that exactly one of the IsXXX functions must return true at all times. 

I don't think that's true

Reply:
Consider adding a test for this for the unstructured implementation, since you removed my tests of invalid values. :)

Reply:
er, now that I see your comment:

* say that that condition makes the value "invalid"
* and put the tests with invalid values back in?